include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+     */
+    MINIMUM_VARIANCE
+};
+
+struct sidb_binary_cluster_hierarchy_node;
+using sidb_binary_cluster_hierarchy_node_ptr = std::unique_ptr<sidb_binary_cluster_hierarchy_node>;
+/**
+ * The struct used to store a binary cluster hierarchy that may be used to store the result of the hierarchical
+ * clustering returned by ALGLIB functionality.
+ */
+struct sidb_binary_cluster_hierarchy_node
+{
+    /**
+     * The set of SiDB indices contained in the node
+     */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+    std::set<uint64_t> c;
+#else
+    phmap::flat_hash_set<uint64_t> c;
+#endif
+    /**
+     * The two children of the node.
+     */
+    std::array<sidb_binary_cluster_hierarchy_node_ptr, 2> sub;
+    /**
+     * Binary cluster hierarchy node constructor.
+     *
+     * @param sidbs The set of SiDBs to contain in this node.
+     * @param children The pair of binary cluster hierarchy node pointers that become the children of this node.
+     */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+    sidb_binary_cluster_hierarchy_node(const std::set<uint64_t>& sidbs,
+#else
+    sidb_binary_cluster_hierarchy_node(phmap::flat_hash_set<uint64_t> sidbs,
+#endif
+                                       std::array<sidb_binary_cluster_hierarchy_node_ptr, 2> children) noexcept :
+            c{std::move(sidbs)},
+            sub{std::move(children)}
+    {}
+};
+/**
+ * This function performs the ALGLIB agglomerative clustering algorithm for a given SiDB layout. By default, the cluster
+ * are created by a minimal positional variance heuristic, also known as Ward's method.
+ *
+ * @tparam Lyt The type of the layout to create a cluster hierarchy of. It must be a cell-level SiDB layout.
+ * @param The The layout to create a cluster hierarchy of.
+ * @param linkage_method The agglomerative clustering linking heuristic that is used by ALGLIB.
+ */
+template <typename Lyt>
+static sidb_binary_cluster_hierarchy_node
+sidb_cluster_hierarchy(Lyt& lyt, sidb_cluster_hierarchy_linkage_method linkage_method =
+                                     sidb_cluster_hierarchy_linkage_method::MINIMUM_VARIANCE) noexcept
+{
+    static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
+    static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
+
+    // no clusterizer call required for <= 1 SiDBs
+    if (lyt.num_cells() == 0)
+    {
+        return sidb_binary_cluster_hierarchy_node{{}, {nullptr, nullptr}};
+    }
+    else if (lyt.num_cells() == 1)
+    {
+        return sidb_binary_cluster_hierarchy_node{{0}, {nullptr, nullptr}};
+    }
+
+    charge_distribution_surface<Lyt> charge_lyt{lyt};
+
+    alglib::real_2d_array d;
+    d.setlength(static_cast<alglib::ae_int_t>(charge_lyt.num_cells()), 2);
+    for (uint64_t i = 0; i < charge_lyt.num_cells(); ++i)
+    {
+        const auto [x, y] = charge_lyt.get_all_sidb_locations_in_nm().at(i);
+
+        d(static_cast<int>(i), 0) = x;
```

Comment:
```suggestion
        d(static_cast<int64_t>(i), 0) = x;
```
is this possible?

Reply:
I think I did this on purpose to deal with ALGLIBs datatypes properly which uses `int`s I believe. Think this avoided some warnings

Reply:
I see. I was expecting that. Thanks for the clarification.